### PR TITLE
🚨🚨🐛 Source Google Analytics: revert `:` auto replacement to underscore in column names

### LINKED ISSUE #30582: GA connector breaking schema change after 1.4.0 version
### Connector Name

source-google-analytics-data-api

### Connector Version

1.4.0

### What step the error happened?

During the sync

### Revelant information

GA4 API allows to fetch custom dimensions by using `customEvent:parameter_name` structure. With `google-analytics-data-api` connector version `1.4.0` a change was introduced where all column symbols (`:`) are being automatically replaced with an underscore (`_`). Related PR: https://github.com/airbytehq/airbyte/pull/30417

After this change we received a breaking schema change notifications for all our GA4 connectors and it seems like there is no way to resolve this from the UI side. We also tried re-creating the connection, but it fails during the sync process in transformation step with error that seems to be related to column not being found (`KeyError: 'customEvent:user_id_dimension'`) when checking for primary key (added additional log output with the traceback). Any ideas how to resolve this?

### Relevant log output

```shell
2023-09-19 14:10:34 destination > 'customEvent:user_id_dimension'
Traceback (most recent call last):
  File "/usr/local/bin/transform-catalog", line 33, in <module>
    sys.exit(load_entry_point('normalization==0.0.0', 'console_scripts', 'transform-catalog')())
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/transform.py", line 111, in main
    TransformCatalog().run(args)
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/transform.py", line 36, in run
    self.process_catalog()
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/transform.py", line 64, in process_catalog
    processor.process(catalog_file=catalog_file, json_column_name=json_col, default_schema=schema)
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/catalog_processor.py", line 80, in process
    nested_processors = stream_processor.process()
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/stream_processor.py", line 298, in process
    self.generate_scd_type_2_model(from_table, column_names),
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/stream_processor.py", line 854, in generate_scd_type_2_model
    "primary_key_partition": self.get_primary_key_partition(column_names),
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/stream_processor.py", line 1059, in get_primary_key_partition
    return [self.get_primary_key_from_path(column_names, path) for path in self.primary_key]
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/stream_processor.py", line 1059, in <listcomp>
    return [self.get_primary_key_from_path(column_names, path) for path in self.primary_key]
  File "/usr/local/lib/python3.7/site-packages/normalization/transform_catalog/stream_processor.py", line 1067, in get_primary_key_from_path
    if "type" in self.properties[field]:
KeyError: 'customEvent:user_id_dimension'
2023-09-19 14:10:34 destination > 
2023-09-19 14:10:34 destination > Showing destination_catalog.json to diagnose/debug errors (1):
2023-09-19 14:10:34 destination > 
2023-09-19 14:10:34 destination > In-connector normalization done (exit code 1)
2023-09-19 14:10:34 replication-orchestrator > (pod: jobs / destination-bigquery-write-4748485-9-fgqkx) - Closed all resources for pod
2023-09-19 14:10:34 replication-orchestrator > Attempt 0 to update stream status incomplete null:events
2023-09-19 14:10:34 replication-orchestrator > Attempt 0 to Flush States from SyncPersistenceImpl
2023-09-19 14:10:35 replication-orchestrator > Attempt 0 to Flush Stats from SyncPersistenceImpl
2023-09-19 14:10:35 replication-orchestrator > Sync worker failed.
java.util.concurrent.ExecutionException: io.airbyte.workers.internal.exception.DestinationException: Destination process exited with non-zero exit code 1
	at java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[?:?]
	at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073) ~[?:?]
	at io.airbyte.workers.general.DefaultReplicationWorker.replicate(DefaultReplicationWorker.java:214) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
	at io.airbyte.workers.general.DefaultReplicationWorker.run(DefaultReplicationWorker.java:153) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
	at io.airbyte.workers.general.DefaultReplicationWorker.run(DefaultReplicationWorker.java:68) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
	at io.airbyte.container_orchestrator.orchestrator.ReplicationJobOrchestrator.runJob(ReplicationJobOrchestrator.java:91) ~[io.airbyte-airbyte-container-orchestrator-dev-8669f25941.jar:?]
	at io.airbyte.container_orchestrator.Application.run(Application.java:78) ~[io.airbyte-airbyte-container-orchestrator-dev-8669f25941.jar:?]
	at io.airbyte.container_orchestrator.Application.main(Application.java:38) ~[io.airbyte-airbyte-container-orchestrator-dev-8669f25941.jar:?]
	Suppressed: io.airbyte.workers.exception.WorkerException: Destination process exit with code 1. This warning is normal if the job was cancelled.
		at io.airbyte.workers.internal.DefaultAirbyteDestination.close(DefaultAirbyteDestination.java:138) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
		at io.airbyte.workers.general.DefaultReplicationWorker.replicate(DefaultReplicationWorker.java:169) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
		at io.airbyte.workers.general.DefaultReplicationWorker.run(DefaultReplicationWorker.java:153) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
		at io.airbyte.workers.general.DefaultReplicationWorker.run(DefaultReplicationWorker.java:68) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
		at io.airbyte.container_orchestrator.orchestrator.ReplicationJobOrchestrator.runJob(ReplicationJobOrchestrator.java:91) ~[io.airbyte-airbyte-container-orchestrator-dev-8669f25941.jar:?]
		at io.airbyte.container_orchestrator.Application.run(Application.java:78) ~[io.airbyte-airbyte-container-orchestrator-dev-8669f25941.jar:?]
		at io.airbyte.container_orchestrator.Application.main(Application.java:38) ~[io.airbyte-airbyte-container-orchestrator-dev-8669f25941.jar:?]
Caused by: io.airbyte.workers.internal.exception.DestinationException: Destination process exited with non-zero exit code 1
	at io.airbyte.workers.general.DefaultReplicationWorker.lambda$readFromDstRunnable$4(DefaultReplicationWorker.java:264) ~[io.airbyte-airbyte-commons-worker-dev-8669f25941.jar:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1589) ~[?:?]
2023-09-19 14:10:35 replication-orchestrator > Could not reliably determine committed record counts, committed record stats will be set to null
2023-09-19 14:10:35 replication-orchestrator > sync summary: {
  "status" : "failed",
  "recordsSynced" : 11175,
  "bytesSynced" : 3089685,
  "startTime" : 1695132590176,
  "endTime" : 1695132635173,
  "totalStats" : {
    "bytesCommitted" : 2050915,
    "bytesEmitted" : 3089685,
    "destinationStateMessagesEmitted" : 12,
    "destinationWriteEndTime" : 0,
    "destinationWriteStartTime" : 1695132590183,
    "meanSecondsBeforeSourceStateMessageEmitted" : 1,
    "maxSecondsBeforeSourceStateMessageEmitted" : 6,
    "recordsEmitted" : 11175,
    "recordsCommitted" : 7144,
    "replicationEndTime" : 0,
    "replicationStartTime" : 1695132590176,
    "sourceReadEndTime" : 1695132609252,
    "sourceReadStartTime" : 1695132595883,
    "sourceStateMessagesEmitted" : 9
  },
  "streamStats" : [ {
    "streamName" : "users",
    "stats" : {
      "bytesCommitted" : 349041,
      "bytesEmitted" : 528468,
      "recordsEmitted" : 2285,
      "recordsCommitted" : 1457
    }
  }, {
    "streamName" : "test_report",
    "stats" : {
      "bytesCommitted" : 425981,
      "bytesEmitted" : 636044,
      "recordsEmitted" : 2316,
      "recordsCommitted" : 1488
    }
  }, {
    "streamName" : "events",
    "stats" : {
      "bytesCommitted" : 1275893,
      "bytesEmitted" : 1925173,
      "recordsEmitted" : 6574,
      "recordsCommitted" : 4199
    }
  } ]
}
2023-09-19 14:10:35 replication-orchestrator > failures: [ {
  "failureOrigin" : "destination",
  "internalMessage" : "Destination process exited with non-zero exit code 1",
  "externalMessage" : "Something went wrong within the destination connector",
  "metadata" : {
    "attemptNumber" : 9,
    "jobId" : 4748485,
    "connector_command" : "write"
  },
  "stacktrace" : "io.airbyte.workers.internal.exception.DestinationException: Destination process exited with non-zero exit code 1\n\tat io.airbyte.workers.general.DefaultReplicationWorker.lambda$readFromDstRunnable$4(DefaultReplicationWorker.java:264)\n\tat java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)\n\tat java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)\n\tat java.base/java.lang.Thread.run(Thread.java:1589)\n",
  "timestamp" : 1695132634690
} ]
```

### Contribute

- [ ] Yes, I want to contribute

### PR DESCRIPTION
**Pull request by lazebnyi**
## What
Some destination does not support auto replacement from `:` auto replacement column name to underscore. 
Fixed https://github.com/airbytehq/airbyte/issues/30582

## How
Revert part of https://github.com/airbytehq/airbyte/pull/30417 what add `:` auto replacement to underscore in column names

## Recommended reading order
1. `x.python`

## 🚨 User Impact 🚨
New users who have synchronised custom reports with dimensions containing symbols using version 1.4.0 - 1.5.0 and are utilizing a destination that does not support replacing symbols with underscores should consider resetting the schema.

1. Pin version 1.5.0 for the [workspace](https://cloud.airbyte.com/workspaces/60644a9a-9cee-48c5-81fd-fd984e738f54/source/1190e94b-57cf-4282-a26b-d939ebea1608) as one of can be affected with revert
2. Deploy version 1.5.1 with revert for others users 
3. Proceed with braking changes as common procedure. 

## Pre-merge Actions
<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py [hunk 1]
```diff
@@ -127,7 +127,7 @@ def test_parse_response(patch_base_class):
             {"name": "totalUsers", "type": "TYPE_INTEGER"},
             {"name": "newUsers", "type": "TYPE_INTEGER"},
             {"name": "sessions", "type": "TYPE_INTEGER"},
-            {"name": "sessionsPerUser:", "type": "TYPE_FLOAT"},
+            {"name": "sessionsPerUser:parameter", "type": "TYPE_FLOAT"},
```
Comment (maxi297) [inline airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py:130]: Why does it now get suffixed by `parameter`? This doesn't seem related to the switch from `_` to `:`

**Review by maxi297 — COMMENTED** [review]
The PR says `New users who have synchronized custom reports with dimensions containing symbols using version 1.4.0 - 1.5.0 and are utilizing a destination that does not support replacing symbols with underscores should consider resetting the schema.`

Therefore, this seems like a breaking change for some people. Do we have the list of those people? Should we send official communication to them?

**Comment by tybernstein** [comment]
Blast radius check shows that no connections had an initial sync on version 1.4.0
Screenshot 2023-09-20 at 12 28 35 PM

**Comment by lazebnyi** [comment]
Can you share link to board where you check this?

File: airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py (hunk 1 above)
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py:130]: Yes, but custom dimentions structure is `<scope>:<parameter>`, so I deside to add parameter after `:`

**Comment by lazebnyi** [comment]
> The PR says `New users who have synchronized custom reports with dimensions containing symbols using version 1.4.0 - 1.5.0 and are utilizing a destination that does not support replacing symbols with underscores should consider resetting the schema.`
> 
> Therefore, this seems like a breaking change for some people. Do we have the list of those people? Should we send official communication to them?

About this I don't know where I can check it. Asked help here - https://airbyte-globallogic.slack.com/archives/C02U9R3AF37/p1695202410243759?thread_ts=1695150542.145859&cid=C02U9R3AF37

**Comment by tybernstein** [comment]
I did not realize there was another version running with this change (1.4.1) [I updated the question HERE](https://airbyte.metabaseapp.com/question/7052-initial-syncs-ran-on-v-1-4-0-of-ga4-source-connector) and there are 13 new connections that were created on v1.4.1.

**Comment by lazebnyi** [comment]
@tybernstein what a state now 
we have bracking chenge wich already merged with this [PR](https://github.com/airbytehq/airbyte/pull/30417)

And now I prepare [revert](https://github.com/airbytehq/airbyte/pull/30608) for them, but as I saw for your [list](https://airbyte-globallogic.slack.com/archives/C02U9R3AF37/p1695202410243759?thread_ts=1695150542.145859&cid=C02U9R3AF37) only 2 workspaces have GAv4 connector version 1.4.0  - 1.5.0 and custom dimentions in custom reports. 

https://cloud.airbyte.com/workspaces/b0d3838d-4bf5-437c-9547-9692f1dc89df/source/9eeb24ec-4342-4ad8-a451-6542c7882599
https://cloud.airbyte.com/workspaces/60644a9a-9cee-48c5-81fd-fd984e738f54/source/1190e94b-57cf-4282-a26b-d939ebea1608

So, my proposition that we can pin version 1.5.0 for them and merge this [PR](https://github.com/airbytehq/airbyte/pull/30608) for other. And prepare PR wich add symbols replacement and proceed with normal bracking changes procedure. 

@maxi297 @brianjlai @tybernstein wdyt?

**Comment by lazebnyi** [comment]
Or we can request oss user reset schema, because any cloud users have not an error and it look like it doesn't affect on any cloud user.

File: airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py (hunk 1 above)
Comment (maxi297) [reply airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py:130]: Is this a breaking change for current users that would have not set up the source after the breaking change as well? If feels like a field is being removed and another one added

**Comment by zypox** [comment]
> @tybernstein The status is now as follows: We have bracking chenge wich already merged with this [PR](https://github.com/airbytehq/airbyte/pull/30417)
> 
> And now I prepare [revert](https://github.com/airbytehq/airbyte/pull/30608) for them, but as I saw for your [list](https://airbyte.metabaseapp.com/question/7052-initial-syncs-ran-on-v-1-4-0-of-ga4-source-connector) only 2 workspaces have GAv4 connector version 1.4.0 - 1.5.0 and custom dimentions in custom reports.
> 
> https://cloud.airbyte.com/workspaces/b0d3838d-4bf5-437c-9547-9692f1dc89df/source/9eeb24ec-4342-4ad8-a451-6542c7882599 https://cloud.airbyte.com/workspaces/60644a9a-9cee-48c5-81fd-fd984e738f54/source/1190e94b-57cf-4282-a26b-d939ebea1608
> 
> So, my proposition that we can pin version 1.5.0 for them and merge this [PR](https://github.com/airbytehq/airbyte/pull/30608) for other. And prepare PR wich add symbols replacement and proceed with normal bracking changes procedure.
> 
> @maxi297 @brianjlai @tybernstein wdyt?

Hey, user here that reported the bug. A few notes:

1. `https://cloud.airbyte.com/workspaces/b0d3838d-4bf5-437c-9547-9692f1dc89df/source/9eeb24ec-4342-4ad8-a451-6542c7882599` is our workspace. We'd like to get new GA connector version with this change reverted, because with current connector version (`1.5.0` ) our syncs are failing and we cannot find a resolution for this.
2. We're a cloud user and we're affected.
3. It was a breaking change, all of our GA connectors that used custom dimensions in them stopped working without us doing anything.

**Comment by zypox** [comment]
> Hey @zypox,
> 
> Yes, we are sorry about the breaking change that was released. Some context, the purpose of this PR is to correct that and return your connections to working properly. Unfortunately, this fix is in turn a breaking change for any Cloud Users that established a connection using one of the broken versions. We need to choose a path forward that minimizes the impact of further breaking changes.
> 
> @lazebnyi your plan makes sense to me

Thanks for your response!

I completely agree with that. I just noticed that @lazebnyi plan includes 1.5.0 version pinning for affected workspaces which includes our workspace as well (`b0d3838d-4bf5-437c-9547-9692f1dc89df`). This would mean that we wouldn't be able to use GA4 source with BigQuery destination (our primary use-case) for as long as we would be pinned to 1.5.0 version. Therefore, I'd like to ask not to pin GA connector version 1.5.0 for the `b0d3838d-4bf5-437c-9547-9692f1dc89df` workspace if it's possible.

**Comment by lazebnyi** [comment]
@zypox as I see your connection successfully synced with 1.5.0 version to bigquery or I miss something?

**Comment by zypox** [comment]
> @zypox as I see your connection successfully synced with 1.5.0 version to bigquery or I miss something?

It synced successfully when we switched from `Normalized tabular data` normalization to `Raw data (JSON)`. With `Normalized tabular data` it's still not working. We switched to raw data so that we could do some manual data analysis while normalised option is not working.

File: airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py (hunk 1 above)
Comment (lazebnyi) [reply airbyte-integrations/connectors/source-google-analytics-data-api/unit_tests/test_streams.py:130]: Should be if they not use normalization in other way the get an error during normalisation.

**Comment by tybernstein** [comment]
Ah @zypox I see, I am sorry I did not notice your workspace was one of the flagged ones. It was included in my query as[ this connection](https://cloud.airbyte.com/workspaces/b0d3838d-4bf5-437c-9547-9692f1dc89df/connections/9ca5ab0f-340e-41e5-bf7c-65ce3638f8ae/job-history) ran its initial sync on 1.5.0 but you would much rather have your older connections working instead of that one, which makes sense.

@lazebnyi I think we are good proceeding with only pinning this connection: https://cloud.airbyte.com/workspaces/60644a9a-9cee-48c5-81fd-fd984e738f54/connections/f733e8f1-7361-485f-b0e5-b8c751cc8f89/job-history

File: airbyte-integrations/connectors/source-google-analytics-data-api/acceptance-test-config.yml [hunk 2]
```diff
@@ -25,32 +25,14 @@ acceptance_tests:
...
-          - name: "ecommerce_purchases_item_category_2_report"
-            bypass_reason: "The test resource does not work with ecommerce purchases"
-          - name: "ecommerce_purchases_item_category_3_report"
-            bypass_reason: "The test resource does not work with ecommerce purchases"
-          - name: "ecommerce_purchases_item_category_4_report"
-            bypass_reason: "The test resource does not work with ecommerce purchases"
-          - name: "ecommerce_purchases_item_category_5_report"
-            bypass_reason: "The test resource does not work with ecommerce purchases"
-          - name: "ecommerce_purchases_item_id_report"
-            bypass_reason: "The test resource does not work with ecommerce purchases"
```
Comment (lazebnyi) [inline airbyte-integrations/connectors/source-google-analytics-data-api/acceptance-test-config.yml:53]: Test resource changed, so some streams can be tested with expected records

**Comment by airbyte-oss-build-runner** [comment]
##  source-google-analytics-data-api test report (commit [`cfbf96d120`](https://github.com/airbytehq/airbyte/pull/30608/commits/cfbf96d1206419d5978e5fddcfa6015a4d686471)) - ✅

⏲️  Total pipeline duration: 52mn14s 

| Step                                                                                    | Result   |
|:----------------------------------------------------------------------------------------|:---------|
| Connector package install                                                               | ✅        |
| Build source-google-analytics-data-api docker image for platform linux/x86_64           | ✅        |
| Unit tests                                                                              | ✅        |
| Acceptance tests                                                                        | ✅        |
| Code format checks                                                                      | ✅        |
| Validate airbyte-integrations/connectors/source-google-analytics-data-api/metadata.yaml | ✅        |
| Connector version semver check                                                          | ✅        |
| Connector version increment check                                                       | ✅        |
| QA checks                                                                               | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/lazebnyi_30582-source-google-analytics-revert-replace-ro-underscore/1695323241/cfbf96d1206419d5978e5fddcfa6015a4d686471/source-google-analytics-data-api/1.5.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:cfbf96d1206419d5978e5fddcfa6015a4d686471)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connectors/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-google-analytics-data-api test
```

**Review by maxi297 — APPROVED** [review]
Given there will be another PR to add symbols replacement and that we confirm we only need to pin one user to 1.5.0, I approve

**Comment by lazebnyi** [comment]
Version 1.5.0 pined for user https://cloud.airbyte.com/workspaces/60644a9a-9cee-48c5-81fd-fd984e738f54/connections/f733e8f1-7361-485f-b0e5-b8c751cc8f89/job-history
